### ISSUE TITLE
Unexpected update in value of `real(dp) :: x(3)`

### ISSUE THREAD

**Issue by Pranavchiku** [lfortran/lfortran#1150]
```fortran
module testmod_der1
private
public fcn, dp
integer, parameter :: dp=kind(0d0)
contains

subroutine fcn(x, fvec)
real(dp), intent(in) :: x(3)
real(dp), intent(out) :: fvec(15)

integer :: i

do i = 1, 3
    print *, "x(1) = ", x(1)
    print *, "x(2) = ", x(2)
    print *, "x(3) = ", x(3)
    fvec(i) = x(1) /( x(2) + x(3) )
    print *, "fvec(i) = ", fvec(i)
end do

end subroutine

end module


program example_lmder1
use testmod_der1, only: dp, fcn
real(dp) :: x(3), fvec(15)
! The following starting values provide a rough fit.
x = [1._dp, 1._dp, 1._dp]

call check_deriv()
contains
subroutine check_deriv()
call fcn(x, fvec)
end subroutine

end program
```

> LFortran
```console
$ lfortran a.f90
x(1) =  1.00000000000000000e+00
x(2) =  1.00000000000000000e+00
x(3) =  1.00000000000000000e+00
fvec(i) =  5.00000000000000000e-01
x(1) =  1.00000000000000000e+00
x(2) =  5.00000000000000000e-01
x(3) =  1.00000000000000000e+00
fvec(i) =  6.66666666666666630e-01
x(1) =  1.00000000000000000e+00
x(2) =  5.00000000000000000e-01
x(3) =  6.66666666666666630e-01
fvec(i) =  8.57142857142857206e-01
```
>GFortran
```console
$ gfortran a.f90 && ./a.out
 x(1) =    1.0000000000000000     
 x(2) =    1.0000000000000000     
 x(3) =    1.0000000000000000     
 fvec(i) =   0.50000000000000000     
 x(1) =    1.0000000000000000     
 x(2) =    1.0000000000000000     
 x(3) =    1.0000000000000000     
 fvec(i) =   0.50000000000000000     
 x(1) =    1.0000000000000000     
 x(2) =    1.0000000000000000     
 x(3) =    1.0000000000000000     
 fvec(i) =   0.50000000000000000 
```

**Comment by Pranavchiku** [comment]
Okay so, if we use the following diff, both GFortran and LFortran produces same results.
```diff
-call check_deriv()
-contains
-subroutine check_deriv()
-call fcn(x, fvec)
-end subroutine
+call fcn(x, fvec)
```

**Comment by Smit-create** [comment]
Thanks for reporting.
I see the bug just by observing the output is that `x(i+1)` also gets updated while updating `fvec(i)`. The global variables are handled using nested_vars pass and have some hackish implementations which need some refactoring.

**Comment by Pranavchiku** [comment]
This error still persists and results in divergence of output of `modern_minpack/example_lmder1.f90`. @certik I guess we used some kind of workaround in legacy minpack, I do not remember it correctly but using that will make `example_lmder1.f90` work.

**Comment by certik** [comment]
Is this a bug in nested functions accessing parent scope variables? 

I can still reproduce it.

In legacy minpakc we don't use nested functions, so we didn't have this problem. Is the ASR correct for this? If so, it's a bug in LLVM, so we need to debug this case and fix it. (The full solution is described here: https://github.com/lfortran/lfortran/issues/1059, but that's for later.)

**Comment by Pranavchiku** [comment]
> Is this a bug in nested functions accessing parent scope variables? 

Yeah. ASR looks good to me. Looking into ASR->LLVM conversion now.

**Comment by Pranavchiku** [comment]
```fortran
module testmod_der1
    private
    public fcn
    contains
    
    subroutine fcn(x, fvec)
    real, intent(in) :: x(3)
    real, intent(out) :: fvec(15)
    
    integer :: i
    
    do i = 1, 3
        print *, "x(1) = ", x(1)
        print *, "x(2) = ", x(2)
        print *, "x(3) = ", x(3)
        fvec(i) = x(1) /( x(2) + x(3) )
        print *, "fvec(i) = ", fvec(i)
    end do
    
    end subroutine
    
    end module
    
    
    program example_lmder1
    use testmod_der1, only: fcn
    real :: x(3), fvec(15)
    ! The following starting values provide a rough fit.
    x = [1.0, 1.0, 1.0]
    
    call check_deriv(x, fvec)
    contains
    subroutine check_deriv(x, fvec)
    real, intent(in) :: x(3)
    real, intent(out) :: fvec(15)
    call fcn(x, fvec)
    end subroutine
    
    end program
```

**Comment by certik** [comment]
Here is a simpler example that also fails:
```
module testmod_der1
private
public fcn, dp
integer, parameter :: dp=kind(0d0)
contains

subroutine fcn(x, fvec)
real(dp), intent(in) :: x
real(dp), intent(out) :: fvec

integer :: i

do i = 1, 3
    print *, "x(1) = ", x
    fvec = x+1
    print *, "fvec(i) = ", fvec
end do

end subroutine

end module


program example_lmder1
use testmod_der1, only: dp, fcn
real(dp) :: x, fvec
! The following starting values provide a rough fit.
x = 1._dp

call check_deriv()
contains
subroutine check_deriv()
call fcn(x, fvec)
end subroutine

end program
```
Here the "x" seems correctly copied to the nested struct, but `fec` is not, so the struct has a dangling pointer in it (or null pointer) and then we access it and it segfaults.

The solution is to copy over a pointer to `fvec` as well.

After we fix the above, we need to fix arrays.

**Comment by certik** [comment]
The following works:
```
program example_lmder1
use testmod_der1, only: dp, fcn
real(dp) :: x, fvec
! The following starting values provide a rough fit.
x = 1._dp
fvec = 2.0

call check_deriv()
contains
subroutine check_deriv()
call fcn(x, fvec)
end subroutine

end program
```
Due to the way nested struct is handled in the assignment. I think we need to disable handling of this in the assignment (at least for this case) and then before calling `check_deriv` we need to copy over the variables into the nested struct, as pointers.

**Comment by Pranavchiku** [comment]
```fortran
program example_lmder1

implicit none
real :: x(3),fvec(15)

! The following starting values provide a rough fit.
x = [1.0, 1.0, 1.0]

call check_deriv()
contains

subroutine check_deriv()
    call fcn(x, fvec)
end subroutine check_deriv

subroutine fcn(x, fvec)

    real, intent(in) :: x(3)
    real, intent(inout) :: fvec(15)

    integer :: i
    
    print *, "in fcn"
    do i = 1, 3
        print *, "x(1) = ", x(1)
        print *, "x(2) = ", x(2)
        print *, "x(3) = ", x(3)
        fvec(i) = (x(1) /(x(2) + x(3)))
        print *, "fvec( ", i, ") = ", fvec(i)
    end do

    end subroutine fcn

end program
```

> GFortran

```console
$ gfortran ./examples/a.f90 && ./a.out
 in fcn
 x(1) =    1.00000000    
 x(2) =    1.00000000    
 x(3) =    1.00000000    
 fvec(            1 ) =   0.500000000    
 x(1) =    1.00000000    
 x(2) =    1.00000000    
 x(3) =    1.00000000    
 fvec(            2 ) =   0.500000000    
 x(1) =    1.00000000    
 x(2) =    1.00000000    
 x(3) =    1.00000000    
 fvec(            3 ) =   0.500000000
```

> LFortran

```console
$ lfortran ./examples/a.f90 
in fcn
x(1) =  1.00000000e+00
x(2) =  1.00000000e+00
x(3) =  1.00000000e+00
fvec(  1 ) =  5.00000000e-01
x(1) =  1.00000000e+00
x(2) =  5.00000000e-01
x(3) =  1.00000000e+00
fvec(  2 ) =  6.66666687e-01
x(1) =  1.00000000e+00
x(2) =  5.00000000e-01
x(3) =  6.66666687e-01
fvec(  3 ) =  8.57142806e-01
```

**Comment by certik** [comment]
@Pranavchiku did you figure out why it happens? We need to fix it.

**Comment by Pranavchiku** [comment]
Yeah, in #1444 I only handled the cases where type of variable is not array. To fix this we need to add provision for case where variabke is array by pointing first element of array in nested struct. I made progress, but it throws me segmentation fault. I will have to debug more on that.

**Comment by certik** [comment]
I see. Post your Draft PR what you have. Maybe I'll see what the issue is.

**Comment by certik** [comment]
Here is a test that does not work: https://github.com/lfortran/lfortran/pull/1457

**Comment by certik** [comment]
This works:
```
module testmod_der1
private
public fcn, dp
integer, parameter :: dp=kind(0d0)
contains

subroutine fcn(x, fvec)
real(dp), intent(in) :: x(3)
real(dp), intent(out) :: fvec(15)

integer :: i

do i = 1, 3
    print *, "x(1) = ", x(1)
    print *, "x(2) = ", x(2)
    print *, "x(3) = ", x(3)
    fvec(i) = x(1) /( x(2) + x(3) )
    print *, "fvec(i) = ", fvec(i)
end do

end subroutine

end module


program example_lmder1
use testmod_der1, only: dp, fcn
real(dp) :: x(3), fvec(15)
! The following starting values provide a rough fit.
x = [1._dp, 1._dp, 1._dp]
fvec = -1

call check_deriv()
contains
subroutine check_deriv()
call fcn(x, fvec)
end subroutine

end program
```
So let's submit a test that checks this.

This does not work yet:
```
module testmod_der1
private
public fcn, dp
integer, parameter :: dp=kind(0d0)
contains

subroutine fcn(x, fvec)
real(dp), intent(in) :: x(3)
real(dp), intent(out) :: fvec(15)

integer :: i

do i = 1, 3
    print *, "x(1) = ", x(1)
    print *, "x(2) = ", x(2)
    print *, "x(3) = ", x(3)
    fvec(i) = x(1) /( x(2) + x(3) )
    print *, "fvec(i) = ", fvec(i)
end do

end subroutine

end module


program example_lmder1
use testmod_der1, only: dp, fcn
real(dp) :: x(3), fvec(15)
! The following starting values provide a rough fit.

fvec = -1
x = [1._dp, 1._dp, 1._dp]

call check_deriv()
contains
subroutine check_deriv()
call fcn(x, fvec)
end subroutine

end program
```
Or if we remove the `fvec = -1` line completely. So this we still need to fix, either in the current design, or if it is too complicated, the new design being implemented in #1464 should take care of this.

**Comment by certik** [comment]
The original issue is fixed:
```console
$ lfortran a.f90 
x(1) =  1.00000000000000000e+00
x(2) =  1.00000000000000000e+00
x(3) =  1.00000000000000000e+00
fvec(i) =  5.00000000000000000e-01
x(1) =  1.00000000000000000e+00
x(2) =  1.00000000000000000e+00
x(3) =  1.00000000000000000e+00
fvec(i) =  5.00000000000000000e-01
x(1) =  1.00000000000000000e+00
x(2) =  1.00000000000000000e+00
x(3) =  1.00000000000000000e+00
fvec(i) =  5.00000000000000000e-01
```
